### PR TITLE
feat: Restrict transfers during lockup period (SC-2045)

### PR DESCRIPTION
**Pull request by michaelelliot**
# Description

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: contracts/Pool.sol [hunk 1]
```diff
@@ -434,7 +434,7 @@ contract Pool is PoolFDT {
     */
     function _transfer(address from, address to, uint256 wad) internal override {
         _whenProtocolNotPaused();
-        PoolLib.prepareTransfer(withdrawCooldown, depositDate, totalCustodyAllowance[from], balanceOf(from), to, wad, _globals(superFactory), balanceOf(to), recognizableLossesOf(from));
+        PoolLib.prepareTransfer(withdrawCooldown, depositDate, totalCustodyAllowance[from], balanceOf(from), from, to, wad, _globals(superFactory), balanceOf(to), recognizableLossesOf(from), lockupPeriod);
```
Comment (lucas-manuel) [inline contracts/Pool.sol:437]: Lets make this multiline now for readability

File: contracts/library/PoolLib.sol [hunk 2]
```diff
@@ -245,16 +245,19 @@ library PoolLib {
...
-        require(recognizableLosses == uint256(0),                "P:RECOG_LOSSES");                  // If an LP has unrecognized losses, they must recognize losses through withdraw
-        updateDepositDate(depositDate, toBalance, wad, to);                                          // Update deposit date of recipient
+        require(depositDate[from].add(lockupPeriod) <= block.timestamp,           "P:FUNDS_LOCKED");                  // Sender must be past lockup period
+        require(fromBalance.sub(wad) >= totalCustodyAllowance,                    "P:INSUFFICENT_TRANSFERABLE_BAL");  // User can only transfer tokens that aren't custodied
+        require(isReceiveAllowed(withdrawCooldown[to], globals),                  "P:RECIPIENT_NOT_ALLOWED");         // Recipient must not be currently withdrawing
+        require(recognizableLosses == uint256(0),                                 "P:RECOG_LOSSES");                  // If an LP has unrecognized losses, they must recognize losses through withdraw
+        updateDepositDate(depositDate, toBalance, wad, to);                                                           // Update deposit date of recipient
```
Comment (lucas-manuel) [inline contracts/library/PoolLib.sol:260]: Too much spacing on error messages

File: contracts/test/StakeLocker.t.sol [hunk 3]
```diff
@@ -280,11 +282,33 @@ contract StakeLockerTest is TestUtil {
...
+        uint256 start = block.timestamp;
+        stakeAmount = constrictToRange(stakeAmount, 1 wei, bPool.balanceOf(address(sam)), true);
+
+        makePublicAndStake(stakeAmount);
+
+        // Will fail because lockup period hasn't passed yet
+        assertTrue(!sam.try_transfer(address(stakeLocker), address(sid), stakeAmount));
+
+        // Warp to after lockup period and transfer
+        hevm.warp(start + stakeLocker.lockupPeriod());
```
Comment (lucas-manuel) [inline contracts/test/StakeLocker.t.sol:295]: We should add another fail assert for `hevm.warp(start + stakeLocker.lockupPeriod() - 1);`

File: contracts/test/PoolLiquidityProvider.t.sol [hunk 4]
```diff
@@ -184,16 +184,19 @@ contract PoolTest is TestUtil {
...
+        uint256 newDeposit = constrictToRange(depositAmt, 50 * USD, depositAmt, true);  // Amount of FDT transferred
 
-        hevm.warp(startDate + 30 days);
+        // Will fail because lockup period hasn't passed yet
+        assertTrue(!liz.try_transfer(address(pool), address(leo), toWad(newDeposit)));
 
-        assertEq(pool.balanceOf(address(leo)), toWad(depositAmt));
-        assertEq(pool.balanceOf(address(liz)), toWad(depositAmt));
+        // Warp to after lockup period
+        hevm.warp(startDate + pool.lockupPeriod());
```
Comment (lucas-manuel) [inline contracts/test/PoolLiquidityProvider.t.sol:199]: We should add another failure assert for `hevm.warp(startDate + pool.lockupPeriod() - 1);`

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Small changes, lets wait to merge this until the interest PR is merged

**Review by lucas-manuel — DISMISSED** [review]
Approving, will merge after https://github.com/maple-labs/maple-core/tree/sc-2046-interest-accrual-bug is in develop

**Review by deluca-mike — DISMISSED** [review]

**Review by deluca-mike — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**